handlers/conversation_handlers.py

```diff
@@ -6,76 +6,115 @@
+    logging.info(f"Пользователь выбрал никнейм: {username}")
+    return states.CONFIRM_EMAIL
+
+
+def change_email(update, context):
+    context.bot.send_message(
+        chat_id=update.effective_chat.id,
+        text=(
+            f"Хорошо, давай поменяем адрес электронной почты.\n"
+            "Пришли адрес, который привязан к твоему аккаунту Notion."
+        ),
+    )
+    logging.info(
+        f"Пользователь {context.user_data[USERNAME]} отправил запрос на смену почты"
+    )
+    return states.CONFIRM_EMAIL
 
 
 def confirmation_sent(update, context):
-    email_address = update.message.text
+    if update.message:
+        email_address = update.message.text
+    else:
+        email_address = context.user_data[EMAIL]
     buttons = [
-        [InlineKeyboardButton(text='Отправить повторно', callback_data='resend_email')],
-        [InlineKeyboardButton(text='Изменить почту', callback_data='change_email_address')]
+        [
+            InlineKeyboardButton(
+                text="Отправить повторно", callback_data=RESEND_CONF_LINK
+            )
+        ],
+        [
+            InlineKeyboardButton(
+                text="Изменить почту", callback_data=UPDATE_EMAIL
+            )
+        ],
     ]
     keyboard = InlineKeyboardMarkup(buttons)
     context.bot.send_message(
         chat_id=update.effective_chat.id,
         text=(
             f"На твою почту {email_address} было отправлено письмо с подтверждением, "
-            "следуй инструкциям в письме для завершения процесса регистрации"),
-        reply_markup=keyboard)
+            "следуй инструкциям в письме для завершения процесса регистрации"
+        ),
+        reply_markup=keyboard,
+    )
+    context.user_data[EMAIL] = email_address
+    logging.info(f"Отправили подтверждение на почту {email_address}")
+    return states.UPDATE_DATA
+
 
+def check_user_data(update, context):
+    action = update.callback_query.data
+    if action == RESEND_CONF_LINK:
+        return confirmation_sent(update, context)
+    elif action == UPDATE_EMAIL:
+        return change_email(update, context)
 
-def edit_details(update, context):
-    answer = update.callback_query.data
-    if answer == 'change_email_address':
-        request_email(update, context)
-    elif answer == 'resend_email':
-        confirmation_sent(update, context)
 
 registration_conv = ConversationHandler(
     entry_points=[
-        CommandHandler('start', register_new_user)
+        CommandHandler("start", register_new_user),
     ],
     states={
-        EMAIL: [
-            MessageHandler(Filters.text, request_email)
+        states.REQUEST_EMAIL: [
+            MessageHandler(Filters.text, request_email),
+        ],
+        states.CONFIRM_EMAIL: [
+            MessageHandler(Filters.text, confirmation_sent),
+        ],
+        states.UPDATE_DATA: [
+            CallbackQueryHandler(
+                check_user_data, pass_user_data=True, pass_chat_data=True
```

Comment:
В документации написано, что  устаревшие. Если честно, я не понял за чем они тут нужно
https://python-telegram-bot.readthedocs.io/en/stable/telegram.ext.callbackqueryhandler.html#telegram-ext-callbackqueryhandler

Reply:
Долго пытался понять эту функцию и наконец дошло зачем ты её такой сделал. Один этот колбэк нужно заменить на  2, а функцию вовсе удалить.
```suggestion
                CallbackQueryHandler(confirmation_sent, pattern=RESEND_CONF_LINK)
                CallbackQueryHandler(change_email, pattern=UPDATE_EMAIL)
```

Reply:
Точно, так и надо было делать с самого начала.
Запутался с этим колбэк квери пока правил все.
Спасибо!